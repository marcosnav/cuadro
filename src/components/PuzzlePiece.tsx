import React, { FC } from 'react';
import * as S from './styled';

interface IProps {
  image: string;
  num: number;
  position: number;
}

/**
 * PuzzlePiece Component
 * {num} is the number in the board where it belongs
 * {position} is the current position of the piece in the board
 * @param props { num: number, position: number }
 */
const PuzzlePiece: FC<IProps> = ({ image, num, position }) => {
  const xAgent = (ref: number) => (ref % 4) * 67;
  const yAgent = (ref: number) => Math.floor(ref / 4) * 67;

  const pieceX = xAgent(position);
  const pieceY = yAgent(position);
  const imageX = -xAgent(num);
  const imageY = -yAgent(num);

  return (
    <S.Piece top={pieceY} left={pieceX}>
      <S.PieceImage image={image} top={imageY} left={imageX} />
    </S.Piece>
  );
};

export default PuzzlePiece;
