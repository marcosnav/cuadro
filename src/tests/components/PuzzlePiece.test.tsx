import { shallow } from 'enzyme';
import React from 'react';
import PuzzlePiece from './../../components/PuzzlePiece';

describe('<PuzzlePiece/>', () => {
  test('renders properly', () => {
    const component = shallow(<PuzzlePiece image={''} num={1} position={3} />);
    expect(component).toMatchSnapshot();
  });

  test('must have a translation based on its position', () => {
    const component = shallow(<PuzzlePiece image={''} num={1} position={3} />);
    expect(component.props().top).toBe(0);
    expect(component.props().left).toBe(201);

    component.setProps({ position: 6 });
    expect(component.props().top).toBe(67);
    expect(component.props().left).toBe(134);

    component.setProps({ position: 9 });
    expect(component.props().top).toBe(134);
    expect(component.props().left).toBe(67);

    component.setProps({ position: 12 });
    expect(component.props().top).toBe(201);
    expect(component.props().left).toBe(0);
  });
});
