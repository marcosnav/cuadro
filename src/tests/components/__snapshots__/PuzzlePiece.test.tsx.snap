// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PuzzlePiece/> renders properly 1`] = `
<Piece
  left={201}
  top={0}
>
  <PieceImage
    image=""
    left={-67}
    top={-0}
  />
</Piece>
`;
